rero_ils/modules/documents/dojson/contrib/unimarctojson/model.py

```diff
@@ -180,50 +169,106 @@ def unimarctoauthor(self, key, value):
-
-    publisher = {}
-    indexes = {}
-    lasttag = '?'
-    for tag in value['__order__']:
-        index = indexes.get(tag, 0)
-        data = value[tag]
-        if type(data) == tuple:
-            data = data[index]
-        if tag == 'a' and index > 0 and lasttag != 'a':
-            publishers.append(publisher)
-            publisher = {}
-        if tag == 'a':
-            place = publisher.get('place', [])
-            place.append(data)
-            publisher['place'] = place
-        elif tag == 'c':
-            name = publisher.get('name', [])
-            name.append(data)
-            publisher['name'] = name
-        elif tag == 'd' and index == 0:
-
-            # 4 digits
-            date = re.match(r'.*?(\d{4})', data).group(1)
-            self['publicationYear'] = int(date)
-
-            # create free form if different
-            if data != str(self['publicationYear']):
-                self['freeFormedPublicationDate'] = data
-        indexes[tag] = index + 1
-        lasttag = tag
-    publishers.append(publisher)
-    return publishers
+def unimarcpublishers_provision_activity_publication(self, key, value):
+    """Get provision activity dates."""
+    def build_place_or_agent_data(code, label, index, add_country):
+        type_per_code = {
+            'a': 'bf:Place',
+            'c': 'bf:Agent'
+        }
+        place_or_agent_data = {
+            'type': type_per_code[code],
+            'label': [{'value': remove_trailing_punctuation(label)}]
+        }
+        if add_country:
+            # country from 102
+            field_102 = unimarctojson.get_fields(tag='102')
+            if field_102:
+                field_102 = field_102[0]
+                country_codes = unimarctojson.get_subfields(field_102, 'a')
+                if country_codes:
+                    place_or_agent_data['country'] = country_codes[0].lower()
+        return place_or_agent_data
+
+    publication = {}
+    ind2 = key[4]
+    type_per_ind2 = {
+        ' ': 'bf:Publication',
+        '_': 'bf:Publication',
+        '0': 'bf:Publication',
+        '1': 'bf:Production',
+        '2': 'bf:Distribution',
+        '3': 'bf:Manufacture'
+    }
+    if ind2 == '4':
+        field_d = value.get('d')
+        if field_d:
+            field_d = force_list(field_d)[0]
+            copyrightDate = self.get('copyrightDate', [])
+            if field_d[0] == 'P':
+                copyrightDate.append('℗ ' + field_d[2:])
+            else:
+                copyrightDate.append('© ' + field_d)
+            self['copyrightDate'] = copyrightDate
+    else:
+        publication = {
+            'type': type_per_ind2[ind2],
+            'statement': [],
+        }
+        subfields_d = utils.force_list(value.get('d'))
+        if subfields_d:
+            subfield_d = subfields_d[0]
+            publication['date'] = subfield_d
+            dates = subfield_d.replace('[', '').replace(']', '').split('-')
+            if dates[0]:
```

Comment:
multiple if can probably be replaced by a `and` operator

Reply:
replaced with `try` `except`